src/app/ClusterStateCache.cpp

```diff
@@ -66,14 +67,58 @@ CHIP_ERROR AttributeCache::UpdateCache(const ConcreteDataAttributePath & aPath,
     return CHIP_NO_ERROR;
 }
 
-void AttributeCache::OnReportBegin()
+CHIP_ERROR ClusterStateCache::UpdateEventCache(const EventHeader & aEventHeader, TLV::TLVReader * apData, const StatusIB * apStatus)
+{
+    if (apData)
+    {
+        //
+        // If we've already seen this event before, there's no more work to be done.
+        //
+        if (aEventHeader.mEventNumber < mHighestReceivedEventNumber)
+        {
+            return CHIP_NO_ERROR;
+        }
+
+        EventData eventData;
+        System::PacketBufferHandle handle;
+        System::PacketBufferTLVWriter writer;
+
+        handle = System::PacketBufferHandle::New(chip::app::kMaxSecureSduLengthBytes);
+
+        writer.Init(std::move(handle), false);
+
+        ReturnErrorOnFailure(writer.CopyElement(TLV::AnonymousTag(), *apData));
+        ReturnErrorOnFailure(writer.Finalize(&handle));
+
+        //
+        // Compact the buffer down to a more reasonably sized packet buffer
+        // if we can.
+        //
+        handle.RightSize();
+
+        eventData.first  = aEventHeader;
+        eventData.second = std::move(handle);
+
+        mEventDataCache.insert(std::move(eventData));
+
+        mHighestReceivedEventNumber = aEventHeader.mEventNumber;
+    }
+    else if (apStatus)
+    {
+        mEventStatusCache[aEventHeader.mPath] = *apStatus;
```

Comment:
Likewise, how should this handle existing event data entries for this path?  This needs to be clearly documented in the API documentation.

Reply:
Updated.